### PR TITLE
Implement Playlist reordering

### PR DESCRIPTION
**Pull request by ColdVolcano**
~~Still a WIP since very long titles/artists break it~~ Not anymore

### CODE REVIEW

File: osu.Game/Overlays/Music/PlaylistList.cs [hunk 1]
```diff
@@ -30,8 +32,15 @@ private void itemSelected(BeatmapSetInfo b)
             OnSelect?.Invoke(b);
         }
 
+        private void beatmapsReordered(IEnumerable<BeatmapSetInfo> b)
```
Comment (jorolf) [inline osu.Game/Overlays/Music/PlaylistList.cs]: any reason to create a method with the only purpose to invoke another action?

Comment (ColdVolcano) [inline osu.Game/Overlays/Music/PlaylistList.cs]: I followed the same as was done in line 30

File: osu.Game/Overlays/Music/PlaylistList.cs [hunk 2]
```diff
@@ -103,6 +114,56 @@ public ItemSearchContainer()
...
+                        ChangeChildDepth(item, childrenList.Count - targetIndex);
+                        hasFindItem = true;
+                    }
+                }
+                TriggerBeatmapSetsChange();
+            }
+
+            public void TriggerBeatmapSetsChange()
+            {
+                beatmapSets = new List<BeatmapSetInfo>();
```
Comment (jorolf) [inline osu.Game/Overlays/Music/PlaylistList.cs]: `beatmapSets = Children.Select(b => b.BeatmapSetInfo).ToList()`?

File: osu.Game/Overlays/Music/PlaylistList.cs [hunk 3]
```diff
@@ -103,6 +111,50 @@ public ItemSearchContainer()
...
+            protected override bool OnMouseMove(InputState state)
+            {
+                mousePosition = state.Mouse.Position;
+                return base.OnMouseMove(state);
+            }
+
+            protected override void Update()
+            {
+                base.Update();
+                PlaylistItem item = Children.FirstOrDefault(i => i.DragStartOffset != Vector2.Zero);
```
Comment (peppy) [inline osu.Game/Overlays/Music/PlaylistList.cs]: this whole method is really ugly

File: osu.Game/Overlays/Music/PlaylistList.cs [hunk 4]
```diff
@@ -103,6 +111,50 @@ public ItemSearchContainer()
...
+                if (item == default(PlaylistItem))
+                    return;
+                List<PlaylistItem> childrenList = Children.ToList();
+                item.MoveTo(new Vector2(0, (mousePosition - item.DragStartOffset).Y));
+
+                int targetIndex = Children.Count(i => i.Position.Y - item.Size.Y / 2 < item.Position.Y - i.Size.Y / 2);
```
Comment (peppy) [inline osu.Game/Overlays/Music/PlaylistList.cs]: this is 99.9999% wrong (you can't just use Size for cases like this)

File: osu.Game/Overlays/Music/PlaylistList.cs [hunk 5]
```diff
@@ -76,12 +81,15 @@ public PlaylistList()
...
         {
-            PlaylistItem itemToRemove = items.Children.FirstOrDefault(item => item.BeatmapSetInfo == beatmapSet);
-            if (itemToRemove != null) items.Remove(itemToRemove);
+            PlaylistItem itemToRemove = items.Children.FirstOrDefault(item => item.BeatmapSetInfo.Hash == beatmapSet.Hash);
+            if (itemToRemove != null)
+                items.Remove(itemToRemove);
+            items.TriggerBeatmapSetsChange();
         }
 
         private class ItemSearchContainer : FillFlowContainer<PlaylistItem>, IHasFilterableChildren
```
Comment (peppy) [inline osu.Game/Overlays/Music/PlaylistList.cs]: i'd rather this was at a framework level as a  `ReorderableFillFlowContainer<T> : FillFlowContainer<T> where T : IDraggable` or something

Comment (peppy) [inline osu.Game/Overlays/Music/PlaylistList.cs]: where's the visualtests?

**Comment by peppy** [comment]
Dragging doesn't work correctly if you drag downwards beyond the end of the list too fast.

File: osu.Game/Overlays/Music/PlaylistList.cs (hunk 4 above)
Comment (Tom94) [inline osu.Game/Overlays/Music/PlaylistList.cs]: Also linear stuff reeeeeally doesn't scale to 100k+ beatmaps. Might be a good idea to just use a `SortedList` and then use `IndexOf`.

**Comment by smoogipooo** [comment]
Closing for the following reasons:
1. Unresolved PR issues
2. Using DragStartOffset is pretty ugly and not the direction we want to go, I think.
3. Should be able to drag outside of the panel and still have it re-order.

**Pull request closed without merging**